cunumeric/array.py

```diff
@@ -955,12 +955,84 @@ def byteswap(self, inplace=False):
 
-    @unimplemented
-    def choose(self, choices, out, mode="raise"):
-        numpy_array = self.__array__(stacklevel=3).choose(
-            choices=choices, out=out, mode=mode
+    def choose(self, choices, out=None, mode="raise", stacklevel=2):
+        a = self
+        if out is not None:
+            out = out.convert_to_cunumeric_ndarray(out)
+
+        if isinstance(choices, list):
+            choices = tuple(choices)
+        is_tuple = isinstance(choices, tuple)
+        if is_tuple:
+            n = len(choices)
+            ch_dtype = choices[0].dtype
+            choices = tuple(
+                self.convert_to_cunumeric_ndarray(choices[i]).astype(ch_dtype)
+                for i in range(n)
+            )
+
+        else:
+            choices = self.convert_to_cunumeric_ndarray(choices)
+            #            if self.ndim > choices.ndim - 1:
+            #                for i in range(self.ndim + 1 - choices.ndim):
+            #                    choices = choices[:, None, ...]
+            n = choices.shape[0]
+            ch_dtype = choices.dtype
+            choices = tuple(
+                self.convert_to_cunumeric_ndarray(choices[i, ...])
+                for i in range(n)
+            )
+
+        if not np.issubdtype(self.dtype, np.integer):
+            raise TypeError("a array should be integer type")
+        if mode == "raise":
+            if (a < 0).any() | (a >= n).any():
+                raise ValueError("invalid entry in choice array")
+        elif mode == "wrap":
+            a = a % n
+        elif mode == "clip":
+            a = a.clip(0, n - 1)
+        else:
+            raise ValueError(
+                f"mode={mode} not understood. Must "
+                "be 'raise', 'wrap', or 'clip'"
+            )
+
+        # we need to broadcast all arrays in choices with
+        # input and output arrays
+        if out is not None:
+            out_shape = broadcast_shapes(a.shape, choices[0].shape, out.shape)
+        else:
+            out_shape = broadcast_shapes(a.shape, choices[0].shape)
+
+        for c in choices:
+            out_shape = broadcast_shapes(out_shape, c.shape)
+
+        # if there is not output specified, create one
+        if out is None:
+            # no output, create one
+            out = ndarray(
+                shape=out_shape,
+                dtype=ch_dtype,
+                stacklevel=stacklevel + 1,
+                inputs=(a, choices),
+            )
+        else:
+            # if output is provided, it shape should be the same as out_shape
+            if out.shape != out_shape:
+                raise ValueError(
+                    f"non-broadcastable output operand with shape "
+                    f" {str(out.shape)}"
+                    f" doesn't match the broadcast shape {str(out_shape)}"
+                )
+
+        ch = tuple(c._thunk for c in choices)  #
+        out._thunk.choose(
```

Comment:
You can only use `out` directly if `out.dtype == ch_dtype`. Otherwise complete the operation as if `out` was not defined, then convert the result. See how it's done in `def mean(...)`.

Note the down-casting behavior:
```
>>> np.choose([0,1,0,1], (np.ones(4), np.zeros(4)))
array([1., 0., 1., 0.]) # dtype is float
>>> x = np.zeros(4, dtype=int)
>>> x
array([0, 0, 0, 0])
>>> np.choose([0,1,0,1], (np.ones(4), np.zeros(4)), out=x)
array([1, 0, 1, 0])
>>> x
array([1, 0, 1, 0]) # dtype is int
```